src/ballet/bmtree/fd_bmtree.h

```diff
@@ -0,0 +1,284 @@
+
+   At a high level, inclusion proofs present a sequence of hash
+   instructions that when executed result in the root commitment.
+
+   Inclusion proof size is O(log n) with regards to tree node count.
+
+   Various types of inclusion proofs exist:
+
+     - Single inclusion proofs (over one leaf node)
+     - Range inclusion proofs (over a contiguous range of leaf nodes)
+     - Sparse inclusion proofs (over an arbitrary subset of leaf nodes) */
+
+#define FD_BMTREE_PREFIX_LEAF   (uchar)0x00
+#define FD_BMTREE_PREFIX_BRANCH (uchar)0x01
+
+/* FD_BMTREE20_NODE_SZ: Size in bytes of a SHA-256-160 tree node */
+#define FD_BMTREE20_NODE_SZ (20UL)
+
+/* FD_BMTREE32_NODE_SZ: Size in bytes of a SHA-256 tree node */
+#define FD_BMTREE32_NODE_SZ (32UL)
+
+/* fd_bmtree20_node_t is the hash of a SHA-256-160 tree node (20 bytes) */
+typedef uchar fd_bmtree20_node_t[FD_BMTREE20_NODE_SZ];
+
+/* fd_bmtree20_node_t is the hash of a SHA-256 tree node (32 bytes) */
+typedef uchar fd_bmtree32_node_t[FD_BMTREE32_NODE_SZ];
+
+/* fd_bmtree*_commit_t computes the root of a binary Merkle tree.
+
+   It requires O(log n) space with regard to the number of nodes.
+
+   During the accumulation phase, the data structure consumes all
+   tree leaf nodes sequentially while calculating and buffering
+   branch nodes of upper layers along the way.
+
+   In the finalization phase, the buffered branch node data is hashed
+   to derive the final root hash.
+
+   The separation of the accumulation and finalization phases is
+   required for trees with leaf counts that are not in `{x | 2^x}`.
+   Those contain at least one branch node with only one child node. */
+
+struct __attribute__((aligned(32))) fd_bmtree20_commit {
+  ulong leaf_idx;
+  ulong leaf_cnt;
+  fd_bmtree20_node_t node_buf[];
+};
+typedef struct fd_bmtree20_commit fd_bmtree20_commit_t;
+
+struct __attribute__((aligned(32))) fd_bmtree32_commit {
+  ulong leaf_idx;
+  ulong leaf_cnt;
+  fd_bmtree32_node_t node_buf[];
+};
+typedef struct fd_bmtree32_commit fd_bmtree32_commit_t;
+
+/* Explanation of the above internal state in fd_bmtree*_commit:
+
+   - `leaf_idx` contains the number of leaf nodes that have been
+     accumulated so far. It is synonymous to the index of within
+     the vector of leaf nodes.
+
+     This is used to check how many branch nodes in the upper layers
+     can be derived with the currently known information.
+
+      The current depth of the layers above the leaf nodes is the
+      number of times the `leaf_idx` is divisible by 2.
+
+   - `leaf_cnt` is the expected number of leaf nodes (constant).
+
+   - `node_buf` is indexed by layer, with 0 being the leaf layer.
+
+     It buffers the left node of the most recently seen pair of nodes
+     in that layer.
+
+     In other words:
+
+     Given a layer `L` containing a vector of nodes known so far,
+     the subset `L_left` contains all nodes with index `i` within
+     that layer where `i%2==0`.
```

Comment:
I initially found this confusing, but then after I understood it, I came up with basically the same explanation you gave.
I tried again and came up with:
"node_buf[L] contains the right-most node in layer L (counting from the bottom) that is a left child of its parent."
Then you can add "More precisely, " and your existing explanation.